### PR TITLE
Magiclysm: Multiple spell effects

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary

```SUMMARY: Mods "Magiclysm: Multiple Spell Effects"```
#### Purpose of change

I have had several requests to make it possible to add multiple effects to spells.
#### Describe the solution

I have added a new member to spell_type which is a vector of `fake_spell`, which has a spell_id, max level, and a bool. This will allow a huge amount of versatility, since what this does is actually cast the new spell, rather than just apply affects. I have added a pair of new spells for stormshaper that illustrate the usage: Lightning Storm basically casts Lightning Blast an additional 3 times after it does its own spell effect.

### CODE REVIEW

**Comment by dkargin** [comment]
I was thinking another approach: to wrap up most of current spell data to a 'spell effect' structure (range, damage, aoe, targets) and change json accordingly. 

Then we could add additional structure 'spell context', which can store/change some data between every effect invocation. This structure should be passed to every effect function. For example, the first effect writes all the targets affected to the list `context.hits`. Next effect 'target_hit_to_damage' converts this number to damage bonus, and writes it to `context.data["damage_bonus"]`. The next effect heals caster for `effect.damage + context.data["damage_bonus"]`. This allows us to introduce some ghetto scripting for spells, (since LUA is gone). 

But generally, it allows us to describe a composition of fx/sfx/text effects in a better way. For example, a single fireball could consist of the following effects:
1. Spawn regular explosion effect with specified radius
2. Spawn sound effect for explosion
3. Apply damage in an area
4. Spawn fd_fire for any target affected

Or some bloody necromancer spell could be described as:
1. Spawn fd_blood around caster
2. Spawn fd_blood around target
3. Do actual spell logic.

Let's return back to current PR. The problem with infinite loops now can be solved by introducing this `context` and by storing current number of invocations there.

**Comment by KorGgenT** [comment]
i don't follow why we need to change everything around in order to do... exactly what everything already can do plus this pr. There's no need for any scripting, since changes are all in C++ for effects, and once this PR is wrapped up modders will have even more freedom to have spells do whatever they want.

File: src/magic.cpp [hunk 1]
```diff
@@ -243,6 +261,13 @@ void spell_type::check_consistency()
             debugmsg( string_format( "ERROR: %s has positive increment and base_casting_time > final_casting_time",
                                      sp_t.id.c_str() ) );
         }
+        if( !sp_t.additional_spells.empty() ) {
```
Comment (BevapDin) [inline src/magic.cpp]: You don't need this check, just iterate over the container. If it's empty, nothing will be done. This check is just redundant.

File: src/magic.cpp [hunk 2]
```diff
@@ -804,6 +829,68 @@ int spell::heal( const tripoint &target ) const
...
+        return false;
+    }
+    return true;
+}
+
+bool spell::cast_all_effects( const tripoint &source, const tripoint &target )
+{
+    // first call the effect of the main spell
+    bool success = cast_spell_effect( source, target );
+    for( fake_spell extra_spell : type->additional_spells ) {
```
Comment (BevapDin) [inline src/magic.cpp]:
```suggestion
    for( const fake_spell &extra_spell : type->additional_spells ) {
```

**Comment by dkargin** [comment]
Right now you can not change spell numbers (damage, area, ...) from another spell/effect without adding a new spell effect in C++, with combined logic. For example, there was an ancient spell `death coil` in warcraft 2, which damages several targets and heals caster for amount equal to damage dealt. Or some modders (me) would like to implement a ritual spell, which consumes/checks some items around and adds proportional values to desired spell effect (area/damage/...).  Right now you should C++ a completely new spell effect for any of these.

Since you've done it this way, that's OK. Spell context can be introduced to current approach of spell chains. And it can deal with infinite loop problem as well.

**Comment by KorGgenT** [comment]
no, you don't need to do a new c++ effect to do new spells with different values, you just need to write up more json for them.

**Comment by foulman** [comment]
I don't want to sound nitpicky, but I would call this spell chaining as that's what it really is, one after another, though with zero casting times on the following spells, which I would probaby do, they could more or less be simultaneous depending on what you were trying to accomplish.

I also don't really see how there's a problem with infinite loops, you'd have to create your own custom spells for that, I'm not going to make any that do, and if the player wants to be stupid and lock up their game with a spell loop, or at least until they run out of mana, that's kinda their choice isn't it?

**Comment by KorGgenT** [comment]
sure, you can call it what you like

as far as infinite loops, if i can try to nip it in the bud i'd rather do that if i can. i think i've tracked down a relatively easy way to do it, too.

**Comment by foulman** [comment]
I would suggest increasing the number of chained spells though, 3 is a bit arbitrary. A hard limit of... say spellcraft rank or spellcraft/2, or maybe just max it out at 10. 
I have wanted to make a 'Minute Meteors' spell for example that this would be perfect for, say every round the player gets another shot. They do need to be able to cancel out of the chain though, with esc perhaps?

**Comment by KorGgenT** [comment]
i never said anything about hard limits. i didn't put in any hard limits.

**Comment by foulman** [comment]
LOL, you're right, I just read too much into the 3 castings of the lightning thing, my bad. so it's unlimited?
Also, are the chained spells succeptible to individual spell failure? Considered to be cast by the caster each time or are they automatic?

**Comment by KorGgenT** [comment]
the reason i haven't called it chained spells is that 
1) you don't "cast" it - it's just a part of the first spell you cast 
2) no, you can't cancel during a spell chain

**Comment by foulman** [comment]
> 
> 
> the reason i haven't called it chained spells is that
> 
>     1. you don't "cast" it - it's just a part of the first spell you cast
> 
>     2. no, you can't cancel during a spell chain
1. That's good.
2. I see that as a small issue then, what if you run out of valid targets?  Such as for scaled up Magic Missile spell that fires off 5 missiles?

**Comment by KorGgenT** [comment]
the monsters don't die until the end of the spell chain

**Comment by foulman** [comment]
Wouldn't that depend on the casting time of the chained spells?  I checked and your lightning blasts each have casting times, so is time passing between them?

**Comment by KorGgenT** [comment]
no, the casting time for lightning blast is just for that spell. spells chained onto another spell ignore casting time, casting cost, and range.

**Comment by foulman** [comment]
Ahhh, okay. That makes sense, but.. I feel like if it's possible you should have it able to go both ways.
For example I immediately thought of a spell chain of attack, phase door, attack, phase door, attack, phase door, etc.. to allow you to safely bounce around the battlefield and still get attacks off. 
I guess I'll need to actually playtest it to get an idea of what can be accomplished and what should be avoided.

**Comment by KorGgenT** [comment]
jenkins rebuild

File: src/magic.cpp [hunk 3]
```diff
@@ -225,6 +243,24 @@ void spell_type::load( JsonObject &jo, const std::string & )
     optional( jo, was_loaded, "casting_time_increment", casting_time_increment, 0.0f );
 }
 
+static bool spell_infinite_loop_check( std::set<spell_id> &spell_effects, const spell_id &sp )
```
Comment (kevingranade) [inline src/magic.cpp]: This check is failing, though it's passing on travis since it isn't loading the mod.
It's adding lightning blast to the spell_effects set, then it *visits* lightning_blast, which fails since it's on the list already.

What you want to do is at each node, only consider the previously visited nodes, this requires you to make copies of your set, but unless someone makes a spell with hundreds of chained effects, that doesn't really matter.

In short, pass spell_effects by copy instead of by reference, and only ever insert the *current* spell effect into the set before recursing to the invoked spell effects.

File: src/magic.cpp [hunk 4]
```diff
@@ -260,6 +298,10 @@ void spell_type::check_consistency()
             debugmsg( string_format( "ERROR: %s has positive increment and base_casting_time > final_casting_time",
                                      sp_t.id.c_str() ) );
         }
+        std::set<spell_id> spell_effect_list;
+        if( spell_infinite_loop_check( spell_effect_list, sp_t.id ) ) {
+            debugmsg( string_format( "ERROR: %s has infinite loop in extra_effects", sp_t.id.c_str() ) );
```
Comment (BevapDin) [inline src/magic.cpp]:
```suggestion
            debugmsg( "ERROR: %s has infinite loop in extra_effects", sp_t.id.c_str() );
```
`debugmsg` calls `string_format` anyway.

Applies as well to other places where `debugmsg` and `string_format` are used together.

**Comment by KorGgenT** [comment]
jenkins rebuild
